src/nvim/buffer.c

```diff
@@ -788,6 +784,12 @@ void buf_freeall(buf_T *buf, int flags)
   bufref_T bufref;
   set_bufref(&bufref, buf);
 
+  // Disable buffer-updates for the current buffer.
+  buf_updates_unload(buf, false);
+  if (!bufref_valid(&bufref)) {
+    // on_detach callback deleted the buffer.
+    return;
+  }
```

Comment:
`on_detach` has textlock set, and like the autocmds below has `b_locked` set here too. Makes it pretty hard to delete the buffer, but it might still be possible; it's at least consistent with below to include the check.